Include/internal/pycore_interpframe.h

```diff
@@ -228,8 +228,11 @@ _PyFrame_SetStackPointer(_PyInterpreterFrame *frame, _PyStackRef *stack_pointer)
  * Frames on the frame stack are incomplete until the
  * first RESUME instruction.
  * Frames owned by a generator are always complete.
+ *
+ * NOTE: We allow racy accesses to the instruction pointer
+ * from other threads for sys._current_frames() and similar APIs.
  */
-static inline bool
+static inline bool _Py_NO_SANITIZE_THREAD
```

Comment:
This function accesses the code object of the frame with a borrowed reference via `_PyFrame_GetCode` so if the reference is gone then this can crash, do we care about that?

Reply:
I think the code object (f_executable) never changes on a `_PyInterpreterFrame`. For the `test_sys.test_current_frames` cases, it's also called under the `PyFrameObject`'s lock. That's not enough for the read of `frame->instr_ptr` below, but it's enough to ensure that the `PyFrameObject`'s `f_frame` pointer won't change.

It would be sufficient to only skip the sanitizer on the `frame->instr_ptr` below, but `_Py_NO_SANITIZE_THREAD` can only be applied at the function level.

Reply:
I am a little concerned about adding `_Py_NO_SANITIZE_THREAD` on this whole function as currently there are data races in `_PyFrame_GetCode` which are suppressed[^1] so if this function  is not sanitized then I think it will hide races in `_PyFrame_GetCode` too.

[^1]: https://github.com/python/cpython/blob/f3bf304c2799c31c045033f22db7eb8766a5f939/Tools/tsan/suppressions_free_threading.txt#L19

Reply:
`_Py_NO_SANITIZE_THREAD` only applies to this function. It doesn't apply to the body of `_PyFrame_GetCode`.

I don't think we have any races involving `_PyFrame_GetCode` anymore, but haven't cleaned up the suppressions file yet. I want to do that in a single pass instead of incrementally, so that it'll be easier to revert in case the TSAN CI becomes flaky.